modules/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/snapshot/SnapshotManagerComponent.java

```diff
@@ -292,71 +291,74 @@ public void exportSnapshot(CoreSession session, String key, OutputStream out) th
+            List<String> reservedKeys) throws IOException, DocumentValidationException {
+        DocumentModel snapDoc = importTmpSnapshot(session, is);
+        validateImportedSnapshot(session, snapDoc.getId(), properties, reservedKeys);
     }
 
     @Override
     public DocumentModel importTmpSnapshot(CoreSession session, InputStream is) throws IOException {
         DocumentModel container = persister.getDistributionRoot(session);
         DocumentRef tmpRef = new PathRef(container.getPathAsString(), IMPORT_TMP);
 
+        // create tmp dir for import
         DocumentModel tmp;
         if (session.exists(tmpRef)) {
             tmp = session.getChild(container.getRef(), IMPORT_TMP);
             session.removeChildren(tmp.getRef());
         } else {
-            tmp = session.createDocumentModel(container.getPathAsString(), IMPORT_TMP, "Workspace");
+            tmp = session.createDocumentModel(container.getPathAsString(), IMPORT_TMP,
+                    DistributionSnapshot.CONTAINER_TYPE_NAME);
             tmp.setPropertyValue(NuxeoArtifact.TITLE_PROPERTY_PATH, "tmpImport");
             tmp = session.createDocument(tmp);
             session.save();
         }
 
+        // run import
         DocumentReader reader = new NuxeoArchiveReader(is);
-        DocumentWriter writer = new SnapshotWriter(session, tmp.getPathAsString());
-
+        DocumentWriter writer = new DocumentModelWriter(session, tmp.getPathAsString()) {
+            @Override
+            protected void beforeCreateDocument(DocumentModel doc) {
+                BaseNuxeoArtifactDocAdapter.fillContextData(doc);
+            }
+        };
         DocumentPipe pipe = new DocumentPipeImpl(10);
         pipe.setReader(reader);
         pipe.setWriter(writer);
         pipe.run();
         reader.close();
         writer.close();
 
-        return session.getChildren(tmp.getRef()).get(0);
+        // make sure it's hidden until validation, and move it
+        DocumentModel snapDoc = session.getChildren(tmp.getRef()).get(0);
+        snapDoc.setPropertyValue(DistributionSnapshot.PROP_HIDE, true);
+        BaseNuxeoArtifactDocAdapter.fillContextData(snapDoc);
+        snapDoc = session.saveDocument(snapDoc);
+        snapDoc = session.move(snapDoc.getRef(), session.getParentDocumentRef(tmp.getRef()), null);
+
+        // cleanup tmp dir
+        session.removeDocument(tmp.getRef());
+        session.save();
+
+        return snapDoc;
+    }
+
+    @Override
+    public void validateImportedSnapshot(CoreSession session, String distribDocId, Map<String, Serializable> properties,
+            List<String> reservedKeys) throws DocumentValidationException {
+        DocumentModel snapDoc = session.getDocument(new IdRef(distribDocId));
+
+        // update props
+        RepositoryDistributionSnapshot repoSnap = (RepositoryDistributionSnapshot) snapDoc.getAdapter(
+                DistributionSnapshot.class);
```

Comment:
What about registering the adapter for `RepositoryDistributionSnapshot` which will avoid the cast here? Otherwise, we can do a one line with `var`:
```suggestion
        var repoSnap = (RepositoryDistributionSnapshot) snapDoc.getAdapter(DistributionSnapshot.class);
```

Reply:
i was tempted to do so but it means adapting tests to account for this new contribution :)
i'll make this change, it's easier.